File: offer46-ba-shu-zi-fan-yi-cheng-zi-fu-chuan-lcof/cpp/Solution.cpp
class Solution {
public:
    int translateNum(int num) {
        if (num <= 9) return 1;
        std::vector<int> digitlist;
        while (num) {
            digitlist.push_back(num % 10);
            num /= 10;
        }
        int n = digitlist.size();
        vector<int> digits(n);
        for (int i = 0; i < n; ++i) {
            digits[i] = digitlist[n - i - 1];
        }
        vector<int> dp(n + 1);
        dp[0] = 1;
        for (int i = 1; i <= n; ++i) {
            dp[i] = dp[i - 1];
            if (i - 2 >= 0 && isValid(digits[i - 2], digits[i - 1])) {
                dp[i] += dp[i - 2];
            }
        }
        return dp[n];
    }
    bool isValid(int a, int b) {
        if (a == 1) return true;
        if (a == 2 && b >= 0 && b <= 5) return true;
        return false;
    }
};

File: offer46-ba-shu-zi-fan-yi-cheng-zi-fu-chuan-lcof/python/Solution.py
class Solution:
    def translateNum(self, num: int) -> int:
        if num <= 9:
            return 1
        digitlist = []
        while num != 0:
            digitlist.append(num % 10)
            num = num // 10
        n = len(digitlist)
        digits = []
        for i in range(n):
            digits.append(digitlist[n-i-1])
        dp = [0] * (n + 1)
        dp[0] = 1
        for i in range(1,n+1):
            dp[i] = dp[i-1]
            if i - 2 >= 0 and self.isValid(digits[i-2], digits[i-1]):
                dp[i] += dp[i-2]
        return dp[n]

    def isValid(self,a,b):
        if a == 1:
            return True
        if a == 2 and 0 <= b <= 5:
            return True
        return False
    